sources/org.osbuild.containers-storage

```diff
@@ -0,0 +1,103 @@
+#!/usr/bin/python3
+"""Provide access to an image in the host system's container storage.
+
+This stage differs from other sources in that the source checks to see if the
+container is available in the host's container storage. This puts a requirement
+on the user to ensure that the container is copied into local storage before
+trying to build an image. The starts by reading the host's
+`/etc/containers/storage.conf` file and then using the config to check if the
+container has been imported.
+
+Unlike all other sources, this source relies on external storage not controlled
+by osbuild itself.
+
+Buildhost commands used: `skopeo`.
+"""
+
+import concurrent.futures
+import hashlib
+import subprocess as sp
+import sys
+
+from osbuild import sources
+from osbuild.util import containers
+
+SCHEMA = """
+"additionalProperties": false,
+"properties": {
+  "items": {
+    "description": "The container image to fetch indexed by the container image id",
+    "type": "object",
+    "additionalProperties": false,
+    "patternProperties": {
+      "sha256:[0-9a-f]{64}": {
+        "type": "object",
+        "additionalProperties": false
+      }
+    }
+  }
+}
+"""
+
+
+class ContainersStorageSource(sources.SourceService):
+
+    content_type = "org.osbuild.containers-storage"
+
+    storage_conf = None
+
+    def local_image_name(self, imagename):
+        """
+        Construct the full image name that references an image with a given checksum in the local storage.
+        """
+        if self.storage_conf is None:
+            conf = containers.get_host_storage()
+        driver = conf["storage"]["driver"]
+        graphroot = conf["storage"]["graphroot"]
+        runroot = conf["storage"]["runroot"]
+
+        return f"containers-storage:[{driver}@{graphroot}+{runroot}]{imagename}"
+
+    def fetch_one(self, checksum, desc) -> None:
+        # Instead of fetching anything, just check that it exists.
+        #
+        # Note that there's an obvious TOCTOU issue here, but it's unavoidable without copying the storage or a
+        # container out of it, which is exactly what we want to avoid with this source.
+        # Unlike all other sources, this source relies on external storage not controlled by osbuild itself.
+        self.exists(checksum, desc)
+
+    def fetch_all(self, items) -> None:
+        # prepare each item as a (checksum, desc) tuple (where desc=None)
```

Comment:
(wondering) I wonder if we should just follow the pattern of the "inline" source here and just do this in a loop (c.f. https://github.com/osbuild/osbuild/blob/main/sources/org.osbuild.inline#L64) as we are just skopeo inspect and that's fast. but total super nitpick, just ignore me (mostly wondering aloud)